### PR TITLE
[gui] Allow tabs to text in GUI

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

This change introduces the support of tabs in texts, which are then correctly displayed in the GUI with up to 4 spaces.

The new text variable for this is `[TABS] tab amount [/TABS]`, or direct use of the `\t`.

The background is to have larger text blocks more structured, as individual spaces are more difficult to lay in lines in the correct vertical order.

## Motivation and context

I once brought it into my binary addon system rework project and now thought of bringing it into master.
Really find it very useful in bigger things as a possibility.

Gosh. As a note to "xbmc/windows/GUIWindowSystemInfo.cpp", the thing really needs to be revised, because it doesn't really fit into international text styles :roll_eyes::smirk:.

## How has this been tested?

See test commit, but can also be direct inside strings.po.

## Screenshots (if appropriate):

Example before:
Bildschirmfoto von 2021-10-03 13-21-38
Bildschirmfoto von 2021-10-03 13-27-14

Example after:
Bildschirmfoto von 2021-10-03 13-20-56
Bildschirmfoto von 2021-10-03 13-31-19
Note about first line, there it takes a bit longer, thats why not done for there :smile: 

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by lrusak** [comment]
I don't really understand the need for `[TABS]`? Can we not just use `\t` or is this for something special?

**Comment by AlwinEsch** [comment]
Thought it is undesirable to take something like this directly in strings.po, since "\n" should also be used as "[CR]", for example.

The small positive with it is that it can be changed more easily with the number, e.g. `[TABS]5[/TABS]`.

But both are possible.

**Comment by garbear** [comment]
I have mixed feelings on tabs. With fixed-width text, tabbing offers the same ability as spacing with less effort because you control the horizontal position (assuming you can control tabbing size, not the case in most apps, but it would be the case in Kodi).

However, for variable-width text, tabbing is unreliable. But inspecting the screenshots I only see tabbing in fixed-width text. Maybe we make fixed-width a requirement for tabbing?

**Comment by AlwinEsch** [comment]
Yes, it has to be used with care and actually only if in strings.po's so that it is adapted to the respective languages.

The only thing I think this is very good is with regard to the TextViewer window from Kodi, as this contains larger things and is partly loaded from files (e.g. with my addon changelog request).

Allowing it definitely brings advantages, and tab use will soon be necessary when my project with Markdown support comes as text. For this purpose, the background has to be brought in a kind of reduced markdown (or as much as possible) in the text viewer.

So that something like this easily possible to add and not with hell of pain to have Kodi's string style :rofl::
Screenshot from 2021-09-06 00-35-29

**Comment by AlwinEsch** [comment]
Here is another background I wanted to bring it to:
Bildschirmfoto vom 2021-01-28 21-50-39

Without having it in order, I found it harder to keep track of.

**Comment by garbear** [comment]
@AlwinEsch concerning the last two pictures you posted,

I think that the first picture is a perfect example for the use of tabs. Code in dialog is an edge case but it can definitely use tabs.

On the other hand, the second shot shows tabs mixed with variable-width font. I think that block alignment is more useful than tab alignment in variable-width situations.

However, the cost of misalignment in variable-width tabbing is low, because even misaligned tabs can be more readable than no tabs, especially when the misalignment is patterned as in your second picture.

File: xbmc/guilib/GUIFontTTF.cpp [hunk 1]
```diff
@@ -1145,3 +1163,9 @@ void CGUIFontTTF::SetGlyphStrength(FT_GlyphSlot slot, int glyphStrength)
   slot->metrics.vertBearingY += dy;
   slot->metrics.vertAdvance  += dy;
 }
+
+float CGUIFontTTF::GetTabSpaceLength()
+{
+  const Character* c = GetCharacter(L'X', 0);
```
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/guilib/GUIFontTTF.cpp:1169]: It's easier to process on fixed-width characters, and there's 2^32 total unicode points, so we just take the space hit and use `uint32_t` for all characters. In this case, both `'X'` ("X") and `L'X'` ("\00X") will expand to a padded "\00\00\00X".

However, thinking of type-correctness, I would perfer `'X'` here. Either way, we cast a unicode sub-page (8 bits or 16 bits) to a full unicode point implicitly. If we were using 16-bit `wchars` for processing, then we could avoid a cast using `L`. But this isn't the case, so we might as well simplify the constant. `L` just adds overhead for new programmers, impacting readability.

**Comment by AlwinEsch** [comment]
If OK, I bring this in. About text viewer it is really good to have, with the others you are right makes it not direct sense, there better to separate parts into different controls (**<>Text view:</><>Value view</>**).

**Review by garbear — APPROVED** [review]
I don't see any more problems

**Pull request merged**